### PR TITLE
React recipe 

### PR DESCRIPTION
**Pull request by damianham**
Add a recipe for a React/Redux SPA using the Preact library.

Depends on https://github.com/amberframework/amber/pull/774

### CODE REVIEW

**Comment by faustinoaq** [comment]
@damianham Take a look to #7 

I think we can add specs for this recipe as well:

```cr
  it "compiles react/preact_redux recipe" do
    generate_app recipe: "react/preact_redux"
    build_app.should be_true
  ensure
    cleanup
  end
```

**Comment by damianham** [comment]
Yes, after if you merge in the outstanding pulls that fix the problems causing the existing recipes to fail the test suite in the order that you mentioned and merge #7 then I can add a spec for this recipe

**Comment by faustinoaq** [comment]
Done! #6 Is fixed by #7 , you can merge master branch here using `git merge master` and add a spec for the react recipe.

File: spec/spec_helper.cr [hunk 1]
```diff
@@ -20,6 +20,10 @@ def build_app
   system("cd app && shards build app")
 end
 
+def test_app
+  system("cd app && crystal spec")
+end
+
```
Comment (faustinoaq) [inline spec/spec_helper.cr]: Nice :+1:

**Comment by faustinoaq** [comment]
@amberframework/contributors This PR still depends on amberframework/amber#774 :sweat_smile:

**Comment by faustinoaq** [comment]
@damianham I just updated the formatting in this recipe  and added more test as well.

BTW, I tried to update redux-router v2 to v4 but there is too many breaking changes, so I gave up :sweat_smile: 

I also commented broken specs. I think this PR would require #15 to pass all Travis tests successfully

Finally I added `# TODO:`, `# HACK:` and `# FIXME: ` comments as workaround to solve pending PRs on amber repo.

@damianham Thanks you for this recipe! :+1:

**Comment by faustinoaq** [comment]
screenshot_20180528_074006

@damianham BTW, I just tried it again and I realize this recipe still requires some additions like delete and read button foreach `Component` with forms.

**Comment by faustinoaq** [comment]
@damianham BTW, Why did you choose `hashHistory` instead of `browserHistory`? I think `browserHistory` has a cleaner look :sweat_smile: WDYT?

Ref: https://github.com/damianham/recipes/blob/830a99693a07d6d63f4350ea94d32bf606d133c6/react/preact_redux/app/src/modules/js/store.js#L2

**Comment by damianham** [comment]
@faustinoaq the delete and read buttons are on the index are they not ?
Also delete and edit are restricted to the object creator

**Comment by faustinoaq** [comment]
> delete and read buttons are on the index are they not ?

@damianham Yeah, I ca see them [here](https://github.com/damianham/recipes/blob/react/react/preact_redux/scaffold/view/src/modules/%7B%7Bname%7D%7D/js/%7B%7Bname%7D%7D_index.js.lqd#L58)

However, I got this:

screenshot_20180528_080124

Maybe the `.map` inside `tbody` is not working as expected :sweat_smile: 

BTW, WDYT about `browserHisotry` (replaces `hashHistory`) ?

**Comment by faustinoaq** [comment]
@damianham `edit` and `delete` are not rendered at all :sweat_smile: 

screenshot_20180528_080640

BTW, If you fix this, please use `npm run release` to minify default bundles before push :sweat_smile:

**Comment by damianham** [comment]
@faustinoaq they are not rendered probably because they are owned by another user.  The {{class_name}}IndexItem renders each row and checks if the user can edit/delete the item with a function called  canEditItem() that is defined in src/lib/js/utils.js.  If the user cannot edit/delete the item only the read button is rendered.

**Comment by faustinoaq** [comment]
@damianham I managed to get `edit` and `delete` button working :+1: 

`user_id` field is missing on generated models :sweat_smile: 

screenshot_20180529_005715

**Comment by faustinoaq** [comment]
WDYT about `hashHisotry` (Using `#`)  vs `browserHistory` (Using HTML5 History API) ?

`http://0.0.0.0:3000/#/categories/new?_k=aovs5g` vs `http://0.0.0.0:3000/categories/new` ?

**Comment by damianham** [comment]
@faustinoaq I imagine that in most web applications where there is an authentication and authorization framework - most objects will be created by a user and thus will probably have a user identity field.  This may not always be the case but I think the behaviour I have added to the preact_redux recipe offers the most common behaviour and can easily be removed or modified by anyone using the recipe to build a SPA web application so I am inclined to leave it as it is.  Not presenting a button for the user to press in the first place is better than subsequently returning an error that they are not allowed press it.  I am OK with a switch to browserHistory.

**Pull request merged**